jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,297 @@
+        }
+    }
+
+    private static final class ExhaustedSubscription implements Flow.Subscription
+    {
+        @Override
+        public void request(long n)
         {
+            // As per rules 3.6 and 3.7, after the Subscription is cancelled all operations MUST be NOPs.
+        }
+
+        @Override
+        public void cancel()
+        {
+            // As per rules 3.6 and 3.7, after the Subscription is cancelled all operations MUST be NOPs.
+        }
+    }
+
+    private static final class ActiveSubscription extends IteratingCallback implements Flow.Subscription
+    {
+        private static final long NO_MORE_DEMAND = -1;
+        private static final Throwable COMPLETED = new StaticException("Source.Content read fully");
+        private final AtomicReference<Throwable> cancelled;
+        private final AtomicLong demand;
+        private Content.Source content;
+        private Flow.Subscriber<? super Content.Chunk> subscriber;
+
+        public ActiveSubscription(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+        {
+            this.cancelled = new AtomicReference<>(null);
+            this.demand = new AtomicLong(0);
             this.content = content;
             this.subscriber = subscriber;
-            this.stalled = true;
         }
 
+        // As per rule 3.3, Subscription MUST place an upper bound on possible synchronous
+        // recursion between Publisher and Subscriber
+        //
+        // As per rule 1.3, onSubscribe, onNext, onError and onComplete signaled to a
+        // Subscriber MUST be signaled serially.
+        //
+        // IteratingCallback guarantee that process() method will be executed by one thread only.
+        // The process() method can be only initiated from request() or cancel() demands methods.
         @Override
-        public void request(long n)
+        protected Action process()
         {
-            boolean process = false;
-            Throwable failure = null;
-            try (AutoLock ignored = lock.lock())
+            Throwable cancelled = this.cancelled.get();
+            if (cancelled != null)
             {
-                if (cancelled || terminated)
-                    return;
-                if (n <= 0)
+                // As per rule 3.13, Subscription.cancel() MUST request the Publisher to eventually
+                // drop any references to the corresponding subscriber.
+                this.demand.set(NO_MORE_DEMAND);
+                if (cancelled != COMPLETED)
+                    this.content.fail(cancelled);
+                this.content = null;
+                try
                 {
-                    terminated = true;
-                    failure = new IllegalArgumentException("invalid demand " + n);
+                    if (cancelled == COMPLETED)
+                        this.subscriber.onComplete();
+                    else if (!(cancelled instanceof Suppressed))
```

Comment:
I don't think this is right. Even in case of `Suppressed` we should notify `onError()`.

Reply:
Suppressed exists explicitly to avoid calling onError. The OP introduced the concept of suppression in order to pass the reactive test suite.